gravatar/src/main/java/com/gravatar/services/AvatarService.kt

```diff
@@ -22,51 +16,35 @@ public class AvatarService(private val okHttpClient: OkHttpClient? = null) {
-    private val coroutineScope = CoroutineScope(GravatarSdkDI.dispatcherMain)
-
     /**
      * Uploads a Gravatar image for the given email address.
      *
      * @param file The image file to upload
      * @param email The email address to associate the image with
      * @param wordpressBearerToken The bearer token for the user's WordPress/Gravatar account
-     * @param gravatarUploadListener The listener to notify of the upload result
      */
-    public fun upload(
-        file: File,
-        email: Email,
-        wordpressBearerToken: String,
-        gravatarUploadListener: GravatarListener<Unit, ErrorType>,
-    ) {
+    public suspend fun upload(file: File, email: Email, wordpressBearerToken: String): Result<Unit, ErrorType> {
         val service = GravatarSdkDI.getGravatarApiV1Service(okHttpClient)
         val identity = MultipartBody.Part.createFormData("account", email.toString())
         val filePart =
             MultipartBody.Part.createFormData("filedata", file.name, file.asRequestBody())
 
-        service.uploadImage("Bearer $wordpressBearerToken", identity, filePart).enqueue(
-            object : Callback<ResponseBody> {
-                override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
-                    coroutineScope.launch {
-                        if (response.isSuccessful) {
-                            gravatarUploadListener.onSuccess(Unit)
-                        } else {
-                            // Log the response body for debugging purposes if the response is not successful
-                            Logger.w(
-                                LOG_TAG,
-                                "Network call unsuccessful trying to upload Gravatar: $response.body",
-                            )
-                            gravatarUploadListener.onError(errorTypeFromHttpCode(response.code()))
-                        }
-                    }
-                }
+        @Suppress("TooGenericExceptionCaught")
+        return try {
+            val response = service.uploadImage("Bearer $wordpressBearerToken", identity, filePart)
 
-                override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
-                    coroutineScope.launch {
-                        gravatarUploadListener.onError(t.error())
-                    }
-                }
-            },
-        )
+            if (response.isSuccessful) {
+                Result.Success(Unit)
+            } else {
+                // Log the response body for debugging purposes if the response is not successful
+                Logger.w(
+                    LOG_TAG,
+                    "Network call unsuccessful trying to upload Gravatar: $response.body",
+                )
+                Result.Failure(errorTypeFromHttpCode(response.code()))
+            }
+        } catch (ex: Exception) {
+            Result.Failure(ex.error())
+        }
```

Comment:
It seems you are catching all exceptions which isn't safe as you're also catching `CancellationException`. Conventional advice is you re-throw cancellation exceptions.

```kotlin
...
return try {
    val response = service.uploadImage("Bearer $wordpressBearerToken", identity, filePart)

    if (response.isSuccessful) {
        Result.Success(Unit)
    } else {
        // Log the response body for debugging purposes if the response is not successful
        Logger.w(
            LOG_TAG,
            "Network call unsuccessful trying to upload Gravatar: $response.body",
        )
        Result.Failure(errorTypeFromHttpCode(response.code()))
    }
} catch( ce: CancelationException) {
    throw ce
} catch (ex: Throwable) {
    Result.Failure(ex.error())
}

```